### PR TITLE
add attributes to dynamic functions

### PR DESCRIPTION
**Pull request by Emilgardis**

### CODE REVIEW

File: src/codegen/mod.rs [hunk 1]
```diff
@@ -3948,9 +3948,13 @@ impl CodeGenerator for Function {
         {
             attributes.push(attributes::must_use());
         }
-
+        let mut comment_ = None;
         if let Some(comment) = item.comment(ctx) {
-            attributes.push(attributes::doc(comment));
+            if ctx.options().dynamic_library_name.is_none() {
+                attributes.push(attributes::doc(comment));
```
Comment (emilio) [inline src/codegen/mod.rs:3954]: So this looks good as a one-off, but it seems to me we should also preserve stuff like the `#[must_use]` attribute, etc.

So perhaps we can do something like this instead:

```
let is_dynamic_function = ctx.options().dynamic_library_name.is_some() && self.kind() == FunctionKind::Function;
```

Then don't push the link_name etc attributes when we're not a dynamic function, and pass the whole attributes array to the dyngen. wdyt?

**Review by emilio — CHANGES_REQUESTED** [review]
This needs a test too, but it looks like a reasonable fix. One suggestion below, since I think we really want more than just the comment.

File: src/codegen/mod.rs (hunk 1 above)
Comment (Emilgardis) [reply src/codegen/mod.rs:3954]: I think that's a good alternative. I'll try to get a test working :+1:

Comment (Emilgardis) [reply src/codegen/mod.rs]: done!

**Comment by Emilgardis** [comment]
after looking at some of the test code (specifically https://github.com/rust-lang/rust-bindgen/blob/f34e4103f304500c96d332f5cecc9067c980d0f9/tests/expectations/tests/comment-indent.rs), looks like rustfmt is ran. And after finding the test-one.sh file, is the diff really checked here? I'm getting a diff

```
--- ./tests/expectations/tests/dynamic_loading_attributes.rs    2022-02-25 20:26:11.144213600 +0100
+++ /tmp/bindings.rs.EAjAoN     2022-02-25 20:43:49.701485812 +0100
@@ -1,8 +1,10 @@
+/* automatically generated by rust-bindgen 0.59.2 */
+
 #![allow(
     dead_code,
     non_snake_case,
     non_camel_case_types,
     non_upper_case_globals
 )]

 extern crate libloading;
@@ -17,37 +19,35 @@
 impl TestLib {
     pub unsafe fn new<P>(path: P) -> Result<Self, ::libloading::Error>
     where
         P: AsRef<::std::ffi::OsStr>,
     {
         let library = ::libloading::Library::new(path)?;
         Self::from_library(library)
     }
-    pub unsafe fn from_library<L>(
-        library: L,
-    ) -> Result<Self, ::libloading::Error>
+    pub unsafe fn from_library<L>(library: L) -> Result<Self, ::libloading::Error>
     where
         L: Into<::libloading::Library>,
     {
         let __library = library.into();
         let foo = __library.get(b"foo\0").map(|sym| *sym)?;
         let baz = __library.get(b"baz\0").map(|sym| *sym)?;
         Ok(TestLib {
             __library,
             foo,
             baz,
         })
     }
     #[must_use]
-    /// @brief A function
-    ///
-    /// @param x
-    /// @param y
-    /// @return int
+    #[doc = " @brief A function"]
+    #[doc = ""]
+    #[doc = " @param x"]
+    #[doc = " @param y"]
+    #[doc = " @return int"]
     pub unsafe fn foo(
         &self,
         x: ::std::os::raw::c_int,
         y: ::std::os::raw::c_int,
     ) -> ::std::os::raw::c_int {
         (self.foo)(x, y)
     }
     pub unsafe fn baz(&self) -> ::std::os::raw::c_int {
```

**Comment by Emilgardis** [comment]
found the source of my confusion, I'm on windows, and `cargo test` in expectations did not work. working on fixing it for windows by disabling the troublesome tests when `target_os="windows"`😅

**Review by emilio — APPROVED** [review]
Looks good to me. We should avoid the clippy needless-late-init warning, but that can be fixed in a follow-up. Thanks for this!

**Pull request merged**